contracts/contracts/token/OUSD.sol

```diff
@@ -445,159 +392,267 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
-        // Set fixed credits per token
-        nonRebasingCreditsPerToken[msg.sender] = _rebasingCreditsPerToken;
+    function _rebaseOptOut(address _account) internal {
+        require(
+            alternativeCreditsPerToken[_account] == 0,
+            "Account must be rebasing"
+        );
+        RebaseOptions state = rebaseState[_account];
+        require(
+            state == RebaseOptions.StdRebasing || state == RebaseOptions.NotSet,
+            "Only standard rebasing accounts can opt out"
+        );
+
+        uint256 oldCredits = creditBalances[_account];
+        uint256 balance = balanceOf(_account);
 
-        // Decrease rebasing credits, total supply remains unchanged so no
-        // adjustment necessary
-        _rebasingCredits = _rebasingCredits.sub(_creditBalances[msg.sender]);
+        // Account
+        rebaseState[_account] = RebaseOptions.StdNonRebasing;
+        alternativeCreditsPerToken[_account] = 1e18;
+        creditBalances[_account] = balance;
+        // Globals
+        _adjustGlobals(-oldCredits.toInt256(), balance.toInt256());
 
-        // Mark explicitly opted out of rebasing
-        rebaseState[msg.sender] = RebaseOptions.OptOut;
-        emit AccountRebasingDisabled(msg.sender);
+        emit AccountRebasingDisabled(_account);
     }
 
     /**
-     * @dev Modify the supply without minting new tokens. This uses a change in
-     *      the exchange rate between "credits" and OUSD tokens to change balances.
+     * @notice Distribute yield to users. This changes the exchange rate
+     *  between "credits" and OUSD tokens to change rebasing user's balances.
      * @param _newTotalSupply New total supply of OUSD.
      */
-    function changeSupply(uint256 _newTotalSupply)
-        external
-        onlyVault
-        nonReentrant
-    {
-        require(_totalSupply > 0, "Cannot increase 0 supply");
+    function changeSupply(uint256 _newTotalSupply) external onlyVault {
+        require(totalSupply > 0, "Cannot increase 0 supply");
 
-        if (_totalSupply == _newTotalSupply) {
+        if (totalSupply == _newTotalSupply) {
             emit TotalSupplyUpdatedHighres(
-                _totalSupply,
-                _rebasingCredits,
-                _rebasingCreditsPerToken
+                totalSupply,
+                rebasingCredits_,
+                rebasingCreditsPerToken_
             );
             return;
         }
 
-        _totalSupply = _newTotalSupply > MAX_SUPPLY
+        totalSupply = _newTotalSupply > MAX_SUPPLY
             ? MAX_SUPPLY
             : _newTotalSupply;
 
-        _rebasingCreditsPerToken = _rebasingCredits.divPrecisely(
-            _totalSupply.sub(nonRebasingSupply)
+        uint256 rebasingSupply = totalSupply - nonRebasingSupply;
+        // round up in the favour of the protocol
+        rebasingCreditsPerToken_ =
+            (rebasingCredits_ * 1e18 + rebasingSupply - 1) /
+            rebasingSupply;
```

Comment:
Extremely unlikely: If 100% of the tokens are nonRebasing the rebasingSupply will be zero and this math operation will fail.

Reply:
Agreed that it's unlikely to happen, but this way allows us to make guarantees that sum(balances) are always smaller than total assets.